### PR TITLE
[None][fix] Stabilize MoE LoRA CUDA graph scratch

### PR DESCRIPTION
**Pull request by achartier**
## Dev Engineer Review

- `moeOp.cpp` initializes complete pinned-host `GemmCoord` arrays and refreshes them when capacity or GEMM shape changes.
- `CudaGraphLoraManager` reserves workspace for the larger of decode capacity and engine-wide `max_num_tokens`.
- `max_num_tokens` now propagates through manager creation and model-engine initialization.
- The change excludes Qwen metadata propagation and FP8 LoRA cache or conversion support.
- CI failures require follow-up investigation.

## QA Engineer Review

- `test_moe_lora_cuda_graph_params.py` adds parameterized coverage for eager-prefill and per-sequence token capacities.
- `test_moe_lora_grouped_gemm.py` adds BF16 CUDA graph replay coverage for shared runners with identical and distinct layer inputs.
- Tests compare replay output with eager output and verify workspace capacity, LoRA rank, and LoRA size.
- Targeted tests and pre-commit checks reportedly pass, but CI failures remain. Coverage is **needs follow-up**.
- No changed integration tests require test-list updates. Existing Qwen MoE LoRA entries remain listed in `test-db/l0_h100.yml` and `qa/llm_function_core.txt`, both with `TIMEOUT` status.

## Per-File QA Perspective

- `cpp/tensorrt_llm/thop/moeOp.cpp`: Verify capture and replay across grouped-GEMM problem-count and dimension changes.
- `tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py`: Verify eager prefill uses engine-wide capacity without reallocating graph-referenced workspace.
- `tensorrt_llm/_torch/pyexecutor/engine/lora.py`: Verify `max_num_tokens` is forwarded unchanged.
- `tensorrt_llm/_torch/pyexecutor/model_engine.py`: Verify engine token capacity reaches manager initialization.
- `cpp/tensorrt_llm/kernels/cutlass_kernels/include/moe_lora_grouped_gemm.h`: Verify callers provide pinned-host arrays with one `GemmCoord` per problem.
- `tensorrt_llm/_torch/moe/fused_moe/fused_moe_cutlass.py`: Verify documentation matches larger-capacity workspace reservation.
- `tests/unittest/_torch/peft/test_moe_lora_cuda_graph_params.py`: Covers workspace reservation for the largest token capacity. No test-list change applies.
- `tests/unittest/_torch/peft/test_moe_lora_grouped_gemm.py`: Covers BF16 shared-runner CUDA graph replay for identical and distinct layer inputs. No test-list change applies.

## Description

Follow-up to #18527 for the remaining dtype-independent routed-expert MoE LoRA CUDA-graph safety issues.

The grouped-GEMM wrappers consume one pinned-host `GemmCoord` upper bound per problem, but the MoE runner allocated and initialized only one entry while passing `problemCount > 1`. Allocate and populate the full per-problem arrays for each LoRA GEMM stage.

The cached MoE runner is also shared by captured decode and eager prefill. Reserve scratch for the larger of the captured decode capacity and the engine-wide token capacity so a later prefill cannot reallocate storage whose addresses were recorded in an existing graph.

This PR does not include Qwen model metadata propagation (merged in #18527) or FP8 LoRA cache/conversion support.

## Test Coverage

- Clean native SM100 build in the tekit container on NVIDIA B200.
- `python3 -m pytest tests/unittest/_torch/peft/test_moe_lora_cuda_graph_params.py tests/unittest/_torch/peft/test_moe_lora_grouped_gemm.py -q` (`16 passed`).
- Added BF16 CUDA-graph replay coverage for multiple MoE layer calls sharing one cached runner, with both shared and distinct inputs/adapters.
- Added unit coverage that eager prefill capacity takes precedence when reserving shared runner scratch.
- Pre-commit hooks on all changed files.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/thop/moeOp.cpp [hunk 1]
```diff
@@ -2069,20 +2069,26 @@ class FusedMoeRunner : public torch::CustomClassHolder
...
+            fill_max_problems(grouped_gemm.fc2.host_max_problem_out_pinned, mLoraDeviceScratchCapacity, 1, hidden_size,
+                lora_max_low_rank);
             if (has_gated)
             {
-                fill_max_problem(grouped_gemm.gated.host_max_problem_in_pinned, 1, lora_max_low_rank, hidden_size);
-                fill_max_problem(grouped_gemm.gated.host_max_problem_out_pinned, 1, inter_size, lora_max_low_rank);
+                fill_max_problems(grouped_gemm.gated.host_max_problem_in_pinned, mLoraDeviceScratchCapacity, 1,
+                    lora_max_low_rank, hidden_size);
+                fill_max_problems(grouped_gemm.gated.host_max_problem_out_pinned, mLoraDeviceScratchCapacity, 1,
+                    inter_size, lora_max_low_rank);
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/thop/moeOp.cpp:2090]: Implemented in b24c55f6132342c89e8acccfb9c54fc54665089f

**Review by xxi-nv — APPROVED** [review]

File: tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py [hunk 2]
```diff
@@ -47,6 +47,7 @@ def __init__(
         overlap_lora_and_base: bool = False,
         device: str = "cuda",
         max_tokens_per_seq: int = 1,
+        max_num_tokens: Optional[int] = None,
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py]: Since `make_cuda_graph_lora_manager` already declares `max_num_tokens: int` as required and the engine always has a value, `Optional[int] = None` plus the `or 0` fallback below silently restores the pre-fix under-reservation for any caller that forgets to pass it. No in-repo caller relies on the default — could this be a required `int` so a missing value fails loudly instead?

File: tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py [hunk 3]
```diff
@@ -112,8 +115,13 @@ def _reserve_moe_lora_workspaces(self, model: torch.nn.Module) -> None:
         happen before any CUDA graph capture so the buffer addresses baked into
         the graph remain valid across replays.
         """
-        # Worst-case tokens in a single captured forward.
-        max_num_tokens = self.max_batch_size * self.max_tokens_per_seq
+        # Captured decode and eager prefill share the same cached runner. Once a
+        # graph records its scratch addresses, a later prefill must not grow
+        # them, so reserve the larger of the two engine-wide limits.
+        max_num_tokens = max(
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py]: The local `max_num_tokens` means "capacity to reserve" while `self.max_num_tokens` means "engine-wide token limit" — same name, different meaning in one method body, and the log line below prints the local one. Suggest renaming the local to `reserve_num_tokens`.

File: tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py [hunk 4]
```diff
@@ -112,8 +115,13 @@ def _reserve_moe_lora_workspaces(self, model: torch.nn.Module) -> None:
...
+            self.max_batch_size * self.max_tokens_per_seq,
+            self.max_num_tokens or 0,
+        )
         reserved = 0
         for _, module in model.named_modules():
             reserve_fn = getattr(module, "reserve_moe_lora_cuda_graph_workspace", None)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py:127]: Not this line, but its callee: `CutlassFusedMoE.reserve_moe_lora_cuda_graph_workspace` still documents `max_num_tokens` as "Worst-case tokens in a captured forward (max_batch_size * max_tokens_per_seq)" (`fused_moe_cutlass.py:456-457`), which is exactly the assumption this PR invalidates. Please update it to state that the value is the larger of the captured-decode capacity and the engine-wide token limit.

File: tests/unittest/_torch/peft/test_moe_lora_grouped_gemm.py [hunk 5]
```diff
@@ -277,6 +277,67 @@ def call():
...
+    num_tokens, hidden_size, inter_size = 4, 128, 256
+    num_experts, top_k, rank = 4, 2, 8
+    num_layers = 4
+
+    layer_inputs = []
+    for layer_idx in range(num_layers):
+        if not distinct_layers and layer_inputs:
+            layer_inputs.append(layer_inputs[0])
+            continue
+        input_seed = 1000 + layer_idx if distinct_layers else 1000
```
Comment (zhaoyangwang-nvidia) [inline tests/unittest/_torch/peft/test_moe_lora_grouped_gemm.py]: The `if distinct_layers else` branches in `input_seed` / `adapter_seed` are unreachable for `layer_idx > 0` because of the `continue` above, so both ternaries always evaluate to the `distinct_layers` form. Dropping them would make the shared-input case easier to read.

File: tests/unittest/_torch/peft/test_moe_lora_cuda_graph_params.py [hunk 6]
```diff
@@ -40,6 +43,44 @@
 _INTER = 256
 
 
+@pytest.mark.parametrize(
+    ("max_batch_size", "max_tokens_per_seq", "max_num_tokens", "expected_tokens"),
+    ((4, 1, 256, 256), (4, 8, 16, 32)),
+)
+def test_workspace_reservation_uses_largest_token_capacity(
```
Comment (zhaoyangwang-nvidia) [inline tests/unittest/_torch/peft/test_moe_lora_cuda_graph_params.py:51]: This test is CPU-only (`device="cpu"`, `CudaGraphLoraParams` patched out) and carries no marker, while the module docstring states these tests require a GPU and every other test in the file has `@requires_cuda`. A one-line note that it is intentionally GPU-free would avoid confusion.

**Review by zhaoyangwang-nvidia — APPROVED** [review]
Approve with nits.

**Review by nv-xtf — APPROVED** [review]
LGTM from disagg side

File: tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py (hunk 2 above)
Comment (achartier) [reply tensorrt_llm/_torch/peft/lora/cuda_graph_lora_manager.py]: Thanks for the review! 4dd0721021de11a58de30eb3af3274d84895938a should address all your comments.

**Review by rosong11 — APPROVED** [review]
